Add test case for matrix multiply-add with large inputs

fangwangcn opened the pull request:
Summary:
- Added test case for addmm, baddbmm and linear with large inputs
- Testing with torch types: float32, float16, bfloat16

Test Plan:
Run unit tests with:
`buck2 run mode/opt //caffe2/test:linalg_re_cuda`

```
...
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok

----------------------------------------------------------------------
Ran 24 tests in 63.224s

OK (skipped=12)
```

Differential Revision: D39718256

zrphercule commented:
Could you please import this diff to fbcode? Thanks!

IvanYashchuk reviewed (changes requested):
It would be better to create a separate test file for matrix multiplications (`test/test_matmul.py`) and also move tests from https://github.com/pytorch/pytorch/pull/85084 therein.
`test/test_linalg.py` was created primarily for testing torch.linalg module and related functionality.
The matmul operation is distinct enough to be put in a separate file.

lezcano commented on test/test_linalg.py:
```diff
@@ -7376,6 +7376,66 @@ def test_permute_matmul(self):
+        name_fn=lambda batch_size, N, M, P: "{}_{}_{}_{}".format(batch_size, N, M, P),
+    )
+    def test_cublas_baddbmm_large_input(self, device, batch_size, N, M, P, dtype):
+        torch.manual_seed(23)
```
This is not needed

lezcano commented on test/test_linalg.py:
```diff
@@ -7376,6 +7376,66 @@ def test_permute_matmul(self):
+        if N == M and M == P:
+            M2_eye = torch.eye(N, device=device, dtype=dtype)
+            out1_eye_gpu = torch.nn.functional.linear(M1, M2_eye.t(), torch.zeros_like(A))
+            self.assertEqual(M1.cpu(), out1_eye_gpu.cpu())
```
Use `M1_cpu` as to avoid an unnecessary copy.

lezcano commented on test/test_linalg.py:
```diff
@@ -7376,6 +7376,66 @@ def test_permute_matmul(self):
+        if N == M and M == P:
+            M2_eye = torch.eye(N, device=device, dtype=dtype).expand(batch_size, N, N)
+            out2_eye_gpu = torch.baddbmm(torch.zeros_like(A), M1, M2_eye, beta=beta, alpha=alpha)
+            self.assertEqual(M1.cpu(), out2_eye_gpu.cpu())
```
Same as above.

lezcano commented on test/test_linalg.py:
```diff
@@ -7376,6 +7376,66 @@ def test_permute_matmul(self):
+
+        # baddbmm
+        alpha, beta = 1.0, 1.0
+        M1 = M1.expand(batch_size, N, M)
+        M2 = M2.expand(batch_size, M, P)
+        A = A.expand(batch_size, N, P)
+        M1_cpu = M1_cpu.expand(batch_size, N, M)
+        M2_cpu = M2_cpu.expand(batch_size, M, P)
+        A_cpu = A_cpu.expand(batch_size, N, P)
```
You may want to do a similar trick as the one above via `map`.

lezcano commented on test/test_linalg.py:
```diff
@@ -7376,6 +7376,66 @@ def test_permute_matmul(self):
+
+        # baddbmm
+        alpha, beta = 1.0, 1.0
+        M1 = M1.expand(batch_size, N, M)
```
Do we want to pass something with a `stride` of zero, or do we want to make it contiguous?

fangwangcn replied:
Hi @lezcano, what do you mean by `stride` of zero? I'm trying to use the same input for the `linear` and `baddbmm`, so that we can do the cross comparison of the results.

lezcano replied:
Then I don't understand why are you parametrising over `batch_size`, rather than simply doing a `unsqueeze(0)`.

fangwangcn replied:
I think it might be better to cover cases when batch_size have different values. If batch_size is always 1, then it's equivalent to unsqueeze(0)

lezcano reviewed (commented):
Left a few comments. Now, I agree with @IvanYashchuk in that it would be great to have a `test_matmul.py` file for matmul-related tests, as the `test_linalg.py` was mostly designed to test the `torch.linalg` module.

IvanYashchuk reviewed (approved):
Changes to `test/test_linalg.py` look good to me.

fangwangcn commented:
Hi @lezcano @IvanYashchuk, the testing keeps fail with the following error:
```
    File "/var/lib/jenkins/workspace/test/test_matmul_cuda.py", line 144, in test_cublas_baddbmm_large_input
      out2_gpu = torch.baddbmm(A, M1, M2, beta=beta, alpha=alpha).cpu()
  RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasGemmStridedBatchedExFix(handle, opa, opb, (int)m, (int)n, (int)k, (void*)&falpha, a, CUDA_R_16BF, (int)lda, stridea, b, CUDA_R_16BF, (int)ldb, strideb, (void*)&fbeta, c, CUDA_R_16BF, (int)ldc, stridec, (int)num_batches, CUDA_R_32F, CUBLAS_GEMM_DEFAULT_TENSOR_OP)`
```
It seems that with GPU device capability lower than 5.3, `bfloat16` type is not supported. I'm adding a conditional filter to bypass cases like this. What do you suggest?

lezcano commented:
According to this
https://github.com/pytorch/pytorch/blob/24adadd4dbcd90b5aba1d4a45847e4ffa83bd6cc/torch/testing/_internal/common_methods_invocations.py#L8512-L8513
`baddbmm` should work with bfloat16 for >= 11.0. Now, it seems that its backward has some issues with SM < 5.3. I'm not sure which one's the right one, but we should probably be consistent across all the tests once we know which one works when.

fangwangcn commented:
Thank you @lezcano, I'll follow that.

fangwangcn commented:
Hi @lezcano, I have signed the EasyCLA following the link above, but the EasyCLA check still failed. How can I trigger a re-check on this?

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@malfet please have a look and LMK what we have to do here?

malfet commented:
@mikekgfb You need to review the PR and also it needs to be rebased against the latest trunk, as currently it can not be merged due to the conflicts

fangwangcn committed:
Add test case for matrix multiply-add with large inputs (#85550)

Summary:
- Added test case for addmm, baddbmm and linear with large inputs
- Testing with torch types: float32, float16, bfloat16

Pull Request resolved: https://github.com/pytorch/pytorch/pull/85550

Test Plan:
Run unit tests with:
`buck2 run mode/opt //caffe2/test:linalg_re_cuda`

```
...
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_1000_10000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_1000_1000_1000_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_float16 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_2_100_100_100_cpu_float32 (test_linalg_re_cuda.TestLinalgReCudaCPU) ... skipped 'Only runs on cuda'
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_10000_10000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_1_10000_1000_10000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_1000_1000_1000_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_bfloat16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_float16 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok
test_addmm_baddbmm_large_input_2_100_100_100_cuda_float32 (test_linalg_re_cuda.TestLinalgReCudaCUDA) ... ok

----------------------------------------------------------------------
Ran 24 tests in 63.224s

OK (skipped=12)
```

Reviewed By: mikekgfb

Differential Revision: D39718256

Pulled By: mikekgfb

fbshipit-source-id: 0f706f067da6ff3e3e0933c3c908e795a5887314

fangwangcn commented:
@malfet Please check whether this is good for merge. Let me know if we need to change anything, thanks!

mikekgfb commented:
@pytorchbot merge
